yjit/src/asm/mod.rs

```diff
@@ -0,0 +1,400 @@
+use std::mem;
+use std::collections::{BTreeMap};
+
+pub mod x86_64;
+
+/// Pointer to a piece of machine code
+/// We may later change this to wrap an u32
+/// Note: there is no NULL constant for CodePtr. You should use Option<CodePtr> instead.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Debug)]
+#[repr(C)]
+pub struct CodePtr(*const u8);
+
+impl CodePtr {
+    pub fn raw_ptr(&self) -> *const u8 {
+        let CodePtr(ptr) = *self;
+        return ptr;
+    }
+
+    fn into_i64(&self) -> i64 {
+        let CodePtr(ptr) = self;
+        *ptr as i64
+    }
+
+    fn into_usize(&self) -> usize {
+        let CodePtr(ptr) = self;
+        *ptr as usize
+    }
+}
+
+impl From<*mut u8> for CodePtr {
```

Comment:
Might it be possible to do `From<NonNull<u8>>` instead? `NonNull<T>` is a `*mut T` that's guaranteed to not be null, and `Option<NonNull<T>>` is guaranteed to be FFI compatible with `*mut T`